gcc 12: struct std::binary_function is deprecated [-Wdeprecated-declarations]

Various files that include Utilities/Array.hpp:
```
In file included from /home/wthrowe/spectre/src/DataStructures/Tensor/Symmetry.hpp:11,
                 from /home/wthrowe/spectre/src/DataStructures/Tensor/TypeAliases.hpp:10,
                 from /home/wthrowe/spectre/src/DataStructures/Tensor/EagerMath/OrthonormalOneform.hpp:8,
                 from /home/wthrowe/spectre/src/DataStructures/Tensor/EagerMath/OrthonormalOneform.cpp:4:
/home/wthrowe/spectre/src/Utilities/Array.hpp:99:21: warning: ‘template<class _Arg1, class _Arg2, class _Result> struct std::binary_function’ is deprecated [-Wdeprecated-declarations]
   99 | struct Equal : std::binary_function<T, T, bool> {
      |                     ^~~~~~~~~~~~~~~
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/12.1.1/include/g++-v12/bits/unique_ptr.h:37,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/12.1.1/include/g++-v12/memory:76,
                 from /home/wthrowe/spectre-gcc-debug-64/SpectrePch.hpp:12:
/usr/lib/gcc/x86_64-pc-linux-gnu/12.1.1/include/g++-v12/bits/stl_function.h:131:12: note: declared here
  131 |     struct binary_function
      |            ^~~~~~~~~~~~~~~
```

--- a/src/Utilities/Array.hpp
+++ b/src/Utilities/Array.hpp
@@ -96,7 +96,7 @@ struct array {
 };
 namespace detail {
 template <typename T = void>
-struct Equal : std::binary_function<T, T, bool> {
+struct Equal {
   constexpr bool inline operator()(const T& lhs, const T& rhs) const {
     return lhs == rhs;
   }

